source/rust_verify/src/verifier.rs

```diff
@@ -1012,8 +1080,37 @@ impl Verifier {
             vir::printer::write_krate(&mut file, &poly_krate);
         }
 
+        let mut before_err_count = 0;
+        if self.args.expand_errors {
+            before_err_count = self.count_errors;
+            self.expand_targets = vec![]; // flush old errors
+            self.expand_flag = false;
+        }
+
         let (time_smt_init, time_smt_run) =
             self.verify_module(compiler, &poly_krate, module, &mut ctx)?;
+
+        // when `expand_errors` flag is set, re-verify this module with split failing assertions to get more precise error message.
+        if self.args.expand_errors
+            && !self.encountered_vir_error
+            && before_err_count < self.count_errors
+        {
+            if module.segments.len() == 0 {
+                compiler
+                    .diagnostic()
+                    .note_without_error("re-verifying root module for error localization");
+            } else {
+                compiler.diagnostic().note_without_error(&format!(
+                    "re-verifying module {} for error localization",
+                    &module_name
+                ));
+            }
+            ctx.debug_expand_targets = self.expand_targets.to_vec();
+            ctx.expand_flag = true;
+            self.expand_flag = true;
+            self.verify_module(compiler, &poly_krate, module, &mut ctx)?;
```

Comment:
It's not clear to me why we reverify the whole module.  I would have thought that we'd reverify just the individual functions that fail, as they fail.  In other words, I would have expected this to be inside the `Create queries to check the validity of proof/exec function bodies` loop.

Reply:
I agree, that sounds better to me too. Thanks!